x/poolmanager/protorev.go

```diff
@@ -4,88 +4,121 @@ import (
+		takerFeeForCommunityPool := sdk.Coin{}
+		if err := takerFeeForCommunityPool.Unmarshal(bz); err == nil {
+			takerFeesForCommunityPool = append(takerFeesForCommunityPool, takerFeeForCommunityPool)
+		}
 	}
-	osmoutils.MustSet(ctx.KVStore(k.storeKey), types.KeyTakerFeeCommunityPoolProtoRev, &newTakerFeeForCommunityPool)
+
+	return takerFeesForCommunityPool
 }
 
-func (k Keeper) GetTakerFeeTrackerForStakers(ctx sdk.Context) (currentTakerFeeForStakers sdk.Coins) {
-	var takerFeeForStakers types.TrackedVolume
-	takerFeeFound, err := osmoutils.Get(ctx.KVStore(k.storeKey), types.KeyTakerFeeStakersProtoRev, &takerFeeForStakers)
-	if err != nil {
-		// We can only encounter an error if a database or serialization errors occurs, so we panic here.
-		// Normally this would be handled by `osmoutils.MustGet`, but since we want to specifically use `osmoutils.Get`,
-		// we also have to manually panic here.
-		panic(err)
+// GetTakerFeeTrackerForCommunityPoolByDenom returns the taker fee for community pool tracker for the specified denom that has been
+// collected since the accounting height.
+func (k Keeper) GetTakerFeeTrackerForCommunityPoolByDenom(ctx sdk.Context, denom string) (sdk.Coin, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyTakerFeeCommunityPoolProtoRevArray)
+	key := types.GetKeyPrefixTakerFeeCommunityPoolProtoRevByDenom(denom)
+
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), nil
 	}
 
-	// If no volume was found, we treat the existing volume as 0.
-	// While we can technically require volume to exist, we would need to store empty coins in state for each pool (past and present),
-	// which is a high storage cost to pay for a weak guardrail.
-	currentTakerFeeForStakers = sdk.Coins(nil)
-	if takerFeeFound {
-		currentTakerFeeForStakers = takerFeeForStakers.Amount
+	takerFeeForCommunityPool := sdk.Coin{}
+	if err := takerFeeForCommunityPool.Unmarshal(bz); err != nil {
+		return sdk.NewCoin(denom, osmomath.ZeroInt()), err
 	}
 
-	return currentTakerFeeForStakers
+	return takerFeeForCommunityPool, nil
 }
 
-func (k Keeper) GetTakerFeeTrackerForCommunityPool(ctx sdk.Context) (currentTakerFeeForCommunityPool sdk.Coins) {
-	var takerFeeForCommunityPool types.TrackedVolume
-	takerFeeFound, err := osmoutils.Get(ctx.KVStore(k.storeKey), types.KeyTakerFeeCommunityPoolProtoRev, &takerFeeForCommunityPool)
-	if err != nil {
-		// We can only encounter an error if a database or serialization errors occurs, so we panic here.
-		// Normally this would be handled by `osmoutils.MustGet`, but since we want to specifically use `osmoutils.Get`,
-		// we also have to manually panic here.
-		panic(err)
-	}
+// UpdateTakerFeeTrackerForCommunityPoolByDenom increases the take fee for community pool tracker for the specified denom by the specified amount.
+func (k Keeper) UpdateTakerFeeTrackerForCommunityPoolByDenom(ctx sdk.Context, denom string, increasedAmt osmomath.Int) error {
```

Comment:
Do we need so many similar methods? What if we only had one that takes store prefix as an input and made short wrappers that would call this shared function?

This would reduce code duplication.

Reply:
Agreed, added the logic to osmoutils here https://github.com/osmosis-labs/osmosis/pull/7240/commits/ee425f046e3a7df31cdd297df7dbb6098cacf834

Lmk if you think I should just keep in poolmanger instead. I thought the logic could be extendable to any other time we need to track coins in the future, which I imagine will occur again.